### ISSUE TITLE
Error occurs while using COPY FROM json file when sequence of the records in a file mismatch

### ISSUE THREAD

**Issue by sayalithange1408** [crate/crate#15184]
### CrateDB version

5.5.0

### CrateDB setup information

Number Of Nodes: 3
CRATE_HEAP_SIZE: 20G / per node
CRATE_JAVA_OPTS: empty

### Problem description

Getting an error - **“Cannot cast value 'column 4 data' to type bigint”** when trying to restore from the json file.( File is attached for reference - 
[my_table_0_04732d9p60sj8e9o60o30c1g.json](https://github.com/crate/crate/files/13656950/my_table_0_04732d9p60sj8e9o60o30c1g.json)  )

However, restore of the same file was successful on the older version of **crate-5.1.0** .

**_Note : The json file is generated with COPY TO command and it is not manipulated at all._**

Here is the link of my post on forum - 
https://community.cratedb.com/t/error-occurs-while-using-copy-from-json-file-when-sequence-of-the-records-in-a-file-mismatch/1668

### Steps to Reproduce

1. Create Table definition - 
       `create table temp.my_table (
	id long,
	first_column long,
	second_column string,
	third_column long,
	fourth_column string,
	fifth_column long,
	sixth_column string,
	seventh_column string ,
	eighth_column TIMESTAMP WITH TIME ZONE,
	nineth_column TIMESTAMP WITH TIME ZONE,
	tenth_column string,
	object_column object as (
      	      sub_column_w long,
     	      sub_column_x ip,
      	      sub_column_y string,
              sub_column_z boolean
    ),
	primary key (eighth_column,fifth_column,id )
) PARTITIONED BY (eighth_column)
WITH (
   "translog.durability" = 'ASYNC'
);`

2. Copy above attached json file to your desired location(e.g `/tmp/crate-bkp/new/`) 
3. Use the command to import from the file -
   `copy temp.my_table FROM '/tmp/crate-bkp/new/my_table_0_04732d9p60sj8e9o60o30c1g.json' RETURN SUMMARY;`

### Actual Result

Error looks something like this on crate console -
![crate-copy-from-error](https://github.com/crate/crate/assets/153707603/720e4bdc-af72-4527-8019-330c3b6ae11b)

### Expected Result

COPY FROM command should work for irregular sequence of records.

**Comment by proddata** [comment]
In earlier versions CrateDB didn't fully validate data types for `COPY FROM/TO`. Until 5.4 this validation could be disabled, however this has been removed with 5.5.

I would suggest to clean/fix your data in your old cluster. Trying to store a string in an integer field, doesn't make much sense and most likely breaks in other steps.

**Comment by amotl** [comment]
Looking at the JSON File sample originally shared at [^1], I can't spot any anomalies which would indicate why it would store a string into an integer field, and I also haven't been able to manipulate the JSONL accordingly to make it work.

Also at [^1], I asked for creating a more minimal reproducer before creating an issue here. Apparently, it hasn't happened like this. The style of the DDL and data structure is confusing.

[^1]: https://community.cratedb.com/t/error-occurs-while-using-copy-from-json-file-when-sequence-of-the-records-in-a-file-mismatch/1668

**Comment by sayalithange** [comment]
@proddata The data in the older cluster never stored a string in an integer field or vice versa. 
Consider a following scenario:
1. Application fires first insert query with columns in the order col1, col2, col3 (i.e. insert into col1, col2, col3....)
2. Subsequently application fires a second query say col2,col3,col1 (i.e. insert into col2, col3, col1....)

Now lets say we want to restore data to newer version 5.5.0. 
So we will take backup using COPY TO command. 
Here generated json file contains column data in the same order used in insert query like below -
    {"col1":1,"col2":38392,"col3":"apple safari",................}
    {"col2":38393,"col3":"apple safari","col1":2,................}

Now while restoring such data to Crate 5.5.0 using COPY FROM command, column data is getting inserted in the same order which is present in JSON file. But it seems that CRATE 5.5.0 is picks up the column sequence to be in specific order(or maybe the order of first record order from JSON). And hence, data type mismatch error is coming for these columns.

**Comment by sayalithange** [comment]
@amotl I probably missed your last statement. Here are the steps with minimal DDL as requested -

1. Create the table on crate-5.1.0 instance and on crate-5.5.0 instance

     create table temp.my_table (
	       	id long,
	       	first_column long,
	       	second_column string,
	       	third_column long,
	        	primary key (id )
      ) 
	WITH (
   	"translog.durability" = 'ASYNC'
	);



2. Run the insert queries with the given order -
	a) insert into temp.my_table (id,first_column,second_column,third_column) values(1,38392,'apple safari',151155);
	b) insert into temp.my_table (id,second_column,third_column,first_column) values(2,'apple safari',23073,31123);

3. Create the backup 
	copy temp.my_table to DIRECTORY '/tmp/cratebackup/';

4. Copy the data to the desired location(e.g /tmp/crate-bkp/) where crate-5.5.0 is installed freshly and restore the backup with COPY FROM.
	copy temp.my_table FROM '/tmp/crate-bkp/my_table_3_.json' RETURN SUMMARY;
	
5. You will get the exception "Cannot cast value 'apple safari' to type 'bigint'"	
	


**In my case, generated json file is something like this -** 
[my_table_3_.json](https://github.com/crate/crate/files/13683262/my_table_3_.json)

**And When I restore this data on my crate-5.5.0 instance I get this error -**
![Screenshot from 2023-12-15 13-53-10](https://github.com/crate/crate/assets/41051888/20b04ff6-55b3-4327-ad7e-e21868698e04)

**Comment by proddata** [comment]
Hi @sayalithange

Thanks I can confirm, that with the **.json** you shared this can be reproduced

```json
{"id":1,"first_column":38392,"second_column":"apple safari","third_column":151155}
{"id":2,"second_column":"apple safari","third_column":23073,"first_column":31123}
```

```sql
create table temp.my_table (
    id long,
    first_column long,
    second_column string,
    third_column long,
    primary key (id)
    );
```

```sql
COPY temp.my_table FROM '/Users/georg/Sandbox/crate-5.5.1/my_table_3_.json' RETURN SUMMARY;
```
leads to

```
+----------------------------------------------------------+----------------------------------------------------------+---------------+-------------+------------------------------------------------------------------------------------------+
| node                                                     | uri                                                      | success_count | error_count | errors                                                                                   |
+----------------------------------------------------------+----------------------------------------------------------+---------------+-------------+------------------------------------------------------------------------------------------+
| {"id": "WAqVX2G4SFiAIdd2VdlTBQ", "name": "Pizzo Camino"} | file:///Users/georg/Sandbox/crate-5.5.1/my_table_3_.json |             1 |           1 | {"Cannot cast value `apple safari` to type `bigint`": {"count": 1, "line_numbers": [2]}} |
+----------------------------------------------------------+----------------------------------------------------------+---------------+-------------+------------------------------------------------------------------------------------------+
```

**Comment by amotl** [comment]
@sayalithange, @proddata: Thank you so much for coming up with a minimal reproducer, and for confirming it.

**Comment by BaurzhanSakhariev** [comment]
Thanks for reporting @sayalithange, it was a regression introduced in 5.5.0.

Fix will be available in the next 5.5.2 hotfix release.
